lib/core/src/Cardano/Wallet/DB/Layer.hs

```diff
@@ -0,0 +1,813 @@
+                    void $ upsert record [ LocalTxSubmissionLastSlot =. sl ]
+
+        , readLocalTxSubmissionPending =
+            fmap (map localTxSubmissionFromEntity)
+            . listPendingLocalTxSubmissionQuery
+
+        , updatePendingTxForExpiry = \wid tip -> ExceptT $
+            selectWallet wid >>= \case
+        Nothing -> pure $ Left $ ErrNoSuchWallet wid
+        Just _ -> Right <$> updatePendingTxForExpiryQuery wid tip
+
+        , removePendingOrExpiredTx = \wid tid -> ExceptT $ do
+            let errNoSuchWallet =
+                    Left $ ErrRemoveTxNoSuchWallet $ ErrNoSuchWallet wid
+            let errNoMorePending =
+                    Left $ ErrRemoveTxAlreadyInLedger tid
+            let errNoSuchTransaction =
+                    Left $ ErrRemoveTxNoSuchTransaction $
+                    ErrNoSuchTransaction wid tid
+            selectWallet wid >>= \case
+                Nothing -> pure errNoSuchWallet
+                Just _  -> selectTxMeta wid tid >>= \case
+                    Nothing -> pure errNoSuchTransaction
+                    Just _ -> do
+                        count <- deletePendingOrExpiredTx wid tid
+                        pure $ if count == 0
+                            then errNoMorePending
+                            else Right ()
+
+        , getTx = \wid tid -> ExceptT $
+            readCheckpoint_ wid >>= \case
+        Nothing -> pure $ Left $ ErrNoSuchWallet wid
+        Just cp -> do
+            metas <- selectTxHistory cp
+                ti wid Nothing W.Descending
+                    [ TxMetaTxId ==. TxId tid ]
+            case metas of
+                [] -> pure (Right Nothing)
+                meta:_ -> pure (Right $ Just meta)
+
+        {-----------------------------------------------------------------------
+                                       Keystore
+        -----------------------------------------------------------------------}
+
+        , putPrivateKey = \wid key -> ExceptT $
+            selectWallet wid >>= \case
+        Nothing -> pure $ Left $ ErrNoSuchWallet wid
+        Just _ -> Right <$> do
+            deleteWhere [PrivateKeyWalletId ==. wid]
+            insert_ (mkPrivateKeyEntity wid key)
+
+        , readPrivateKey = selectPrivateKey
+
+        {-----------------------------------------------------------------------
+                                 Blockchain Parameters
+        -----------------------------------------------------------------------}
+
+        , readGenesisParameters = selectGenesisParameters
+
+        {-----------------------------------------------------------------------
+                                 Delegation Rewards
+        -----------------------------------------------------------------------}
+
+        , putDelegationRewardBalance =
+            \wid amt -> ExceptT $
+            selectWallet wid >>= \case
+        Nothing -> pure $ Left $ ErrNoSuchWallet wid
+        Just _  -> Right <$> repsert
```

Comment:
Similar to my previous comment!

```suggestion
            selectWallet wid >>= \case
                Nothing -> pure $ Left $ ErrNoSuchWallet wid
                Just _  -> Right <$> repsert
```

Reply:
strangely enough the module is a trove of that *cases*. I tried to stay in style with it, even if I do not like it.